### PR TITLE
[validator] Disable Explicit Class Name Group

### PR DESCRIPTION
**Pull request by zerustech**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |
## Issue:

If "class name" group is **explicity** configured in a constraint, the
constraint will not be added to "Default" group.

The following example reproduces this issue:

``` php
class Foo
{
    /**
     * @Assert\NotNull(
     *     groups = {"Foo"}
     * )
     */
    public $nameInFooGroup;

    /**
     * @Assert\NotNull
     */
    public $nameWithoutGroup;

    /**
     * @Assert\NotNull(
     *     groups = {"Default"}
     * )
     */
    public $nameInDefaultGroup;

}
```

Constraints in "Foo" Group:
1.  `NotNull` constraint on `nameWithoutGroup` property
2.  `NotNull` constraint on `nameInDefaultGroup` property
3.  `NotNull` constraint on `nameInFooGroup` property

Constraints in "Default" Group:
1.  `NotNull` constraint on `nameWithoutGroup` property
2.  `NotNull` constraint on `nameInDefaultGroup` property

The `NotNull` constraint on `nameInFooGroup` property is not added to "Default" group.
This is incorrect because "Default" group and "class name" group should be identical
if neither "group sequence" nor "embed object" is used.
## Cause of the Issue:

The implementation of "class name" group didn't follow the JSR-303 spec.
### JSR-303 Spec:

For class X, group X can not be explicitly defined, because all explicit groups
are implemented as inner interfaces of class X and none of the interfaces can use X
as its name, because `class X is already defined`.
This explains why, in the formal group definition, group X does not include constraints
explicitly assigned to group X, because that's impossible in java.

> The group X contains the following constraints:
> - all Default constraints hosted on X
> - all Default constraints hosted on interfaces of X
> - ...
### Symfony Implementation:

This is the official documentation for "Default" group.

> Constraints in the Default group of a class are the constraints that have either no
> explicit group configured or that are **configured to a group equal to the class name** or
> the string Default.

It does not clarify the meaning of "**configured to a group equal to the class name**":
does it mean "implicitly configured" or "explicitly configured"?

If the class name group is not explicitly defined, then everything is fine. Unfortunately,
in Symfony, it's also possible to explicitly configure the "class name" group for constraints:

``` php
...
class User
{
    /**
     * @Assert\NotNull(
     *      groups = {"User"}
     * )
    */
    public $name;
    ...
}
```

In this case, the `NotNull` constraint on `name` property won't be added to Default group
and that's the issue I am trying to fix.
## Solution:

Follow the JSR-303 spec to forbid the configuration of explicit "class name" group and
when an explicit class name group is found, a `GroupDefinitionException` exception shall be thrown.

### CODE REVIEW

**Comment by zerustech** [comment]
@WouterJ @sstok @webmozart @stof Please check, thanks.

**Comment by webmozart** [comment]
Hi @zerustech, thank you for creating this PR! Why would you add a constraint to the class group? Or is your idea simply to prevent doing this by mistake?

**Comment by zerustech** [comment]
Hi @webmozart, Thanks for your feedback. I think we should prevent developers from doing this, because the symfony document didn't clarify whether or not the "class name group" can be explicitly specified and so developers may make mistakes.

**Comment by webmozart** [comment]
Sounds like a good idea then.

File: src/Symfony/Component/Validator/Constraint.php [hunk 1]
```diff
@@ -231,10 +232,16 @@ public function __get($option)
...
      *
      * @api
+     *
+     * @throws GroupDefinitionException If the implicit group has been explicitly configured in current constraint.
      */
     public function addImplicitGroupName($group)
     {
-        if (in_array(self::DEFAULT_GROUP, $this->groups) && !in_array($group, $this->groups)) {
+        if (in_array($group, $this->groups)) {
+            throw new GroupDefinitionException(sprintf('The implicit group "%s" has already been explicitly configured in constraint %s', $group, get_class($this)));
```
Comment (webmozart) [inline src/Symfony/Component/Validator/Constraint.php]: Although this error is technically correct, it isn't very helpful. I would reword the message so that developers hitting the error have the most benefit. They will hit the error when they define the implicit group name themselves, so they should remove it or use some other group instead. This error message should reflect that IMO.

File: src/Symfony/Component/Validator/Mapping/ClassMetadata.php [hunk 2]
```diff
@@ -217,7 +217,24 @@ public function getDefaultGroup()
     }
 
     /**
-     * {@inheritdoc}
+     * Adds a class constraint.
```
Comment (webmozart) [inline src/Symfony/Component/Validator/Mapping/ClassMetadata.php]: Why did you inline the parent description here?

File: src/Symfony/Component/Validator/Mapping/ClassMetadata.php [hunk 3]
```diff
@@ -262,6 +279,10 @@ public function addConstraint(Constraint $constraint)
      * @param Constraint $constraint The constraint
      *
      * @return ClassMetadata This object
+     *
+     * @throws GroupDefinitionException When the "class name" group has
```
Comment (webmozart) [inline src/Symfony/Component/Validator/Mapping/ClassMetadata.php]: I'd replace `"class name" group` by `default group`, since this is what `getDefaultGroup()` returns.

**Comment by webmozart** [comment]
Status: Needs work

File: src/Symfony/Component/Validator/Constraint.php (hunk 1 above)
Comment (zerustech) [inline src/Symfony/Component/Validator/Constraint.php]: Thanks for your feedback. Since I'm not an English native speaker, could you advice the rewording?

File: src/Symfony/Component/Validator/Mapping/ClassMetadata.php (hunk 2 above)
Comment (zerustech) [inline src/Symfony/Component/Validator/Mapping/ClassMetadata.php]: Actually, the comment for ClassMetadata is not exactly the same as it for the GenericMetadata because for class meta data, its traversal strategy is affected by a `Traverse` constraint, while for generic meta data, `Valid` constraint should be used.

File: src/Symfony/Component/Validator/Mapping/ClassMetadata.php (hunk 3 above)
Comment (zerustech) [inline src/Symfony/Component/Validator/Mapping/ClassMetadata.php]: I think it makes sense to use `"class name"` group because the `"Default"` group can be explicitly defined without any problems ( please refer to the code example in my original post ), this issue only occurs when the `"class name"` group is explicitly defined.

Comment (webmozart) [inline src/Symfony/Component/Validator/Mapping/ClassMetadata.php]: "default group" is the name used in the comments to refer to what you call "class name group". Also, `getDefaultGroup()` returns that group name. In the end, "Default" is just an alias of the default group (i.e. the "class name group").

Comment (zerustech) [inline src/Symfony/Component/Validator/Mapping/ClassMetadata.php]: If we use the word "default group" in the comment, it will confuse the users. Because they can't tell if we are talking about the "Default" group or the "<ClassName>" group. 
Technically, they are supposed to be identical to each other, but the fact is, if the user explicitly defines "ClassName" group, the issue occurs.

**Comment by fabpot** [comment]
Closing this old PR as there is no more activity. Feel free to rebase and reopen.
